File: core/greeter.py
import socket
from config import Config
from logger import LogSender
from watcher import Petter

def Serve(config):
    petter = Petter(config, 'greeter')
    log = LogSender(config, 'greeter')

    # create udp server
    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    addr = ('0.0.0.0', config.GreeterPort())
    sock.bind(addr)

    log.Send('Server started: ' + config.VersionStr() +
             ', UDP: ' + str(addr))

    try:
        while True:
            # respond to broadcast with config str
            data, addr = sock.recvfrom(4096)
            sock.sendto(config.Str().encode('utf-8'), addr)
            log.Send(str(addr), 1)
    except(KeyboardInterrupt):
        pass

    log.Send('Server stopped')
    petter.Stop()

if __name__ == '__main__':
    c = Config()
    Serve(c)

File: core/lib/greeter.js
/*
 * Greeter: Discovery and config server
 *
 * Copyright (c) Shaoqing Tan. All rights reserved.
 */

const dgram = require('dgram');

class Greeter {
  constructor({ config, logger }, cb) {
    // remember payload
    this.data = JSON.stringify(config);
    this.logger = logger || console;

    // answer udp requests (the client will use a broadcast to survey)
    this.udp = dgram.createSocket('udp4');

    // call cb when we ready
    this.udp.on('listening', () => {
      this.logger.log(`Server started: ${config.VersionStr}, UDP: ${config.GreeterPort}`);
      if (cb) {
        cb();
      }
    });

    // receive config of device as json
    this.udp.on('message', (message, remote) => {
      this.logger.log(`${remote.address}:${remote.port}`);
      // ignore errors, we can't really do anything
      this.udp.send(this.data, remote.port, remote.address);
    });

    this.udp.bind({ port: config.GreeterPort });
  }

  close(cb) {
    this.udp.close();
    this.logger.log('Server stopped');
    if (cb) {
      cb();
    }
  }
}

module.exports = Greeter;
